screeps-game-api/src/objects/impls/room_terrain.rs

```diff
@@ -1,15 +1,37 @@
-use {constants::Terrain, objects::RoomTerrain};
+use stdweb::UnsafeTypedArray;
+
+use {constants::{ReturnCode, Terrain}, objects::RoomTerrain, traits::TryInto};
 
 impl RoomTerrain {
     pub fn constructor(room_name: &str) -> Self {
-        js_unwrap!{new Room.Terrain(@{room_name})}
+        js_unwrap!(new Room.Terrain(@{room_name}))
     }
 
     pub fn get(&self, x: u32, y: u32) -> Terrain {
-        js_unwrap!{@{self.as_ref()}.get(@{x}, @{y})}
+        js_unwrap!(@{self.as_ref()}.get(@{x}, @{y}))
     }
 
     pub fn get_raw_buffer(&self) -> Vec<u8> {
-        js_unwrap!(@{self.as_ref()}.getRawBuffer())
+        let mut buffer: Vec<u8> = vec![0; 2500];
+        self.get_raw_buffer_array_ref(&mut buffer).expect("Panic in get_raw_buffer.");
+        buffer
     }
+
+    pub fn get_raw_buffer_array_ref<'a>(&self, buffer: &'a mut Vec<u8>) -> Result<&'a mut Vec<u8>, ReturnCode> {
+        let is_success: bool;
+        {
+            let arr: UnsafeTypedArray<u8> = unsafe{ UnsafeTypedArray::new(buffer.as_mut_slice()) };
+
+            is_success = js!{
+                var bytes = @{arr};
+                return @{self.as_ref()}.getRawBuffer(bytes) != -10;
```

Comment:
Unrelated to other concerns:

Do you think it would be reasonable to `assert_eq!(bytes.len(), 2500);` before we call into JS or should we assume that `getRawBuffer` will error if we give it an incorrect length?

Reply:
I was going with the second. The error code should be exactly for that purpose, so checking here would be duplicating the effort.

Reply:
On the other hand, we could check in our function and panic on any error as they're not supposed to happen.